From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Woobeen Jeon <woobeen0516@gmail.com>
Date: Wed, 7 Jan 2026 12:04:28 +0900
Subject: [PATCH] Minimize TPS lag


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 62ae421128a0238b09ec6ad746daff7178bd5d4c..604f0100356dfd0bc9805005789288529888abc7 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1615,6 +1615,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.server.spark.tickStart(); // Paper - spark
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
+        org.wrd.chrona.tick.TickCalculator.tick(); // Chrona - Minimize TPS lag
         this.tickCount++;
         this.tickRateManager.tick();
         this.tickChildren(hasTimeLeft);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index dc65503a2d785d64d37b76b0303f51cf66d9769a..af12ec029249b10fc82952b383b21d151124d075 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -890,6 +890,10 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     protected void tickTime() {
         if (this.tickTime) {
             long l = this.levelData.getGameTime() + 1L;
+            // Chrona start - Minimize TPS lag
+            if(Boolean.TRUE.equals(this.server.chronaConfiguration.optimization.minimizeTPSLag.getValue()))
+                l += org.wrd.chrona.tick.TickCalculator.getMissedTicks();
+            // Chrona end - Minimize TPS lag
             this.serverLevelData.setGameTime(l);
             Profiler.get().push("scheduledFunctions");
             this.serverLevelData.getScheduledEvents().tick(this.server, l);
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 8463c43e7128d529c71bf6828cdf868661c65f6e..cc6ba1f81b4defd00a86f95259ca195e77511922 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -514,6 +514,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
 
         this.tickEffects();
+        // Chrona start - Minimize TPS lag
+        if(Boolean.TRUE.equals(this.level().getServer().chronaConfiguration.optimization.minimizeTPSLag.getValue())) {
+            for (int i = 0; i < org.wrd.chrona.tick.TickCalculator.getMissedTicks(); i++)
+                this.tickEffects();
+        }
+        // Chrona end - Minimize TPS lag
         this.yHeadRotO = this.yHeadRot;
         this.yBodyRotO = this.yBodyRot;
         this.yRotO = this.getYRot();
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index e82bb8f2d2b0e771dc371fd7e709116e1f5f204f..287c25e8b3a767463b26bd228f827eb79a00959d 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -123,6 +123,10 @@ public class ItemEntity extends Entity implements TraceableEntity {
         super.inactiveTick();
         if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
             this.pickupDelay--;
+            // Chrona start - Minimize TPS lag
+            if(Boolean.TRUE.equals(this.level().getServer().chronaConfiguration.optimization.minimizeTPSLag.getValue()))
+                this.pickupDelay -= org.wrd.chrona.tick.TickCalculator.getMissedTicks();
+            // Chrona end - Minimize TPS lag
         }
         if (this.age != -32768) {
             this.age++;
@@ -148,6 +152,10 @@ public class ItemEntity extends Entity implements TraceableEntity {
             super.tick();
             if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
                 this.pickupDelay--;
+                // Chrona start - Minimize TPS lag
+                if(Boolean.TRUE.equals(this.level().getServer().chronaConfiguration.optimization.minimizeTPSLag.getValue()))
+                    this.pickupDelay -= org.wrd.chrona.tick.TickCalculator.getMissedTicks();
+                // Chrona end - Minimize TPS lag
             }
 
             this.xo = this.getX();
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 7bf160cef8bd772d27d4a71f04d5d9861ef67984..417bc60dc9b5f67e0fa4e8348e7edec71471dfeb 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -256,6 +256,10 @@ public abstract class Player extends Avatar implements ContainerUser {
 
         if (this.isSleeping()) {
             this.sleepCounter++;
+            // Chrona start - Minimize TPS lag
+            if(Boolean.TRUE.equals(this.level().getServer().chronaConfiguration.optimization.minimizeTPSLag.getValue()))
+                this.sleepCounter += org.wrd.chrona.tick.TickCalculator.getMissedTicks();
+            // Chrona end - Minimize TPS lag
             // Paper start - Add PlayerDeepSleepEvent
             if (this.sleepCounter == SLEEP_DURATION) {
                 if (!new io.papermc.paper.event.player.PlayerDeepSleepEvent((org.bukkit.entity.Player) getBukkitEntity()).callEvent()) {
@@ -273,6 +277,10 @@ public abstract class Player extends Avatar implements ContainerUser {
             }
         } else if (this.sleepCounter > 0) {
             this.sleepCounter++;
+            // Chrona start - Minimize TPS lag
+            if(Boolean.TRUE.equals(this.getServer().chronaConfiguration.optimization.minimizeTPSLag.getValue()))
+                this.sleepCounter += org.wrd.chrona.tick.TickCalculator.getMissedTicks();
+            // Chrona end - Minimize TPS lag
             if (this.sleepCounter >= 110) {
                 this.sleepCounter = 0;
             }
