From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Woobeen Jeon <woobeen0516@gmail.com>
Date: Thu, 8 Jan 2026 20:04:49 +0900
Subject: [PATCH] Http API


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 604f0100356dfd0bc9805005789288529888abc7..18adab263c9b2ea5565437e82d9b8b7c40ae545e 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -307,6 +307,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public boolean isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
     private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
+    // Chrona start - Http server
+    @org.jetbrains.annotations.Nullable
+    public org.wrd.chrona.util.Threaded<org.wrd.chrona.http.ApiServer> httpServer;
+    // Chrona end - Http server
     // Paper start - improve tick loop
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes1s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes5s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
@@ -1018,6 +1022,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
+        try {
+            if(this.httpServer != null)
+                this.httpServer.stop(); // Chrona - Http server
+        } catch (Exception e) {
+            LOGGER.error("Failed to stop http server", e);
+        }
         // CraftBukkit start
         if (this.server != null) {
             this.server.spark.disable(); // Paper - spark
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index b0356aa2a9d791f7e201ae48be3db7c87ea9b751..95b3c8eeb6fa9783368302bd79154d60341db5f3 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -317,6 +317,17 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             return false;
         }
 
+        // Chrona start - Http server
+        if(Boolean.TRUE.equals(this.chronaConfiguration.httpServer.enabled.getValue())) {
+            this.httpServer = org.wrd.chrona.util.Threaded.wrap(new org.wrd.chrona.http.ApiServer(
+                    org.wrd.chrona.http.ServerProperties.builder()
+                            .port(this.chronaConfiguration.httpServer.port.getValue())
+                            .build()
+            ));
+            this.httpServer.start();
+        }
+        // Chrona end - Http server
+
         // CraftBukkit start
         this.server.loadPlugins();
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
